firewood/src/storage/mod.rs

```diff
@@ -78,78 +79,19 @@ impl Ash {
     }
 }
 
-#[derive(Debug)]
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
 pub struct AshRecord(pub HashMap<SpaceId, Ash>);
 
 impl growthring::wal::Record for AshRecord {
     fn serialize(&self) -> growthring::wal::WalBytes {
-        let len_bytes = (self.0.len() as u64).to_le_bytes();
-        let ash_record_bytes = self
-            .0
-            .iter()
-            .map(|(space_id, w)| {
-                let space_id_bytes = space_id.to_le_bytes();
-                let undo_len = u32::try_from(w.undo.len()).expect("size of undo shoud be a `u32`");
-                let ash_bytes = w.iter().flat_map(|(undo, redo)| {
-                    let undo_data_len = u64::try_from(undo.data.len())
-                        .expect("length of undo data shoud be a `u64`");
-
-                    undo.offset
-                        .to_le_bytes()
-                        .into_iter()
-                        .chain(undo_data_len.to_le_bytes())
-                        .chain(undo.data.iter().copied())
-                        .chain(redo.data.iter().copied())
-                });
-
-                (space_id_bytes, undo_len, ash_bytes)
-            })
-            .flat_map(|(space_id_bytes, undo_len, ash_bytes)| {
-                space_id_bytes
-                    .into_iter()
-                    .chain(undo_len.to_le_bytes())
-                    .chain(ash_bytes)
-            });
-
-        len_bytes.into_iter().chain(ash_record_bytes).collect()
+        bincode::serialize(self).unwrap().into()
```

Comment:
We should just inline this. Types that implement Serialize shouldn't also have a serialize method.